src/env.rs

```diff
@@ -0,0 +1,278 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An environment which is passed downward into the widget tree.
+
+// TODO: remove this
+#![allow(unused)]
+
+use std::collections::HashMap;
+use std::ops::{Deref, Index};
+
+use crate::piet::Color;
+
+use crate::Data;
+
+// Discussion question: since druid is pretty rigorously single-threaded,
+// maybe we should use Rc instead of Arc.
+use std::sync::Arc;
+
+/// An environment passed down through all widget traversals.
+///
+/// All widget methods have access to an environment, and it is passed
+/// downwards during traversals.
+///
+/// A widget can retrieve theme parameters (colors, dimensions, etc.). In
+/// addition, it can pass custom data down to all descendants. An important
+/// example of the latter is setting a value for enabled/disabled status
+/// so that an entire subtree can be disabled ("grayed out") with one
+/// setting.
+#[derive(Clone)]
+pub struct Env(Arc<EnvImpl>);
+
+struct EnvImpl {
+    theme: EnvValue,
```

Comment:
Something to think about with theme is how dynamic it is. That's to say: for system widgets, we will know all of the recognized theme keys at compile time, so `theme` could just be a struct. This would let us do things like `theme.text_color` and `theme.selected_text_bg_color`, without having to mess around with keys; and we could also guarantee that all values were present in the theme.

There are other options. We could have something like,

```rust
pub struct Theme(EnvMap);

impl Theme {
    fn text_color(&self) -> Color {
            self.0.theme_value(TEXT_COLOR_KEY).as_color()
    }
}
```

which lets us keep using `EnvValue` under the covers but gives a simpler API for this specific (theme) case. (I'm not sure I like this much.)